File: swift/Sources/Shield/RatchetSession.swift
import Foundation
import CommonCrypto

/// RatchetSession provides forward secrecy through key ratcheting.
public class RatchetSession {
    private var sendKey: [UInt8]
    private var recvKey: [UInt8]
    private var sendCounter: UInt64 = 0
    private var recvCounter: UInt64 = 0
    private let isInitiator: Bool

    public init(rootKey: [UInt8], isInitiator: Bool) throws {
        guard rootKey.count == Shield.keySize else {
            throw ShieldError.invalidKeySize
        }

        self.isInitiator = isInitiator

        if isInitiator {
            self.sendKey = RatchetSession.deriveChainKey(rootKey, info: "init_send")
            self.recvKey = RatchetSession.deriveChainKey(rootKey, info: "init_recv")
        } else {
            self.sendKey = RatchetSession.deriveChainKey(rootKey, info: "init_recv")
            self.recvKey = RatchetSession.deriveChainKey(rootKey, info: "init_send")
        }
    }

    public func encrypt(_ plaintext: [UInt8]) throws -> [UInt8] {
        let messageKey = RatchetSession.deriveChainKey(sendKey, info: "message")

        guard let nonce = Shield.randomBytes(Shield.nonceSize) else {
            throw ShieldError.randomGenerationFailed
        }

        // Generate keystream and XOR
        let keystream = generateKeystream(key: messageKey, nonce: nonce, length: plaintext.count)
        var ciphertext = [UInt8](repeating: 0, count: plaintext.count)
        for i in 0..<plaintext.count {
            ciphertext[i] = plaintext[i] ^ keystream[i]
        }

        // Counter bytes
        let counterBytes = withUnsafeBytes(of: sendCounter.littleEndian) { Array($0) }

        // MAC over counter || nonce || ciphertext
        let macData = counterBytes + nonce + ciphertext
        let mac = Shield.hmacSha256(key: messageKey, data: macData)

        // Ratchet
        sendKey = RatchetSession.deriveChainKey(sendKey, info: "ratchet")
        sendCounter += 1

        // Format: counter(8) || nonce(16) || ciphertext || mac(16)
        return counterBytes + nonce + ciphertext + Array(mac.prefix(Shield.macSize))
    }

    public func decrypt(_ encrypted: [UInt8]) throws -> [UInt8] {
        guard encrypted.count >= 8 + Shield.nonceSize + Shield.macSize else {
            throw ShieldError.ciphertextTooShort
        }

        // Parse
        let counter = encrypted.prefix(8).withUnsafeBytes { $0.load(as: UInt64.self).littleEndian }
        let nonce = Array(encrypted[8..<(8 + Shield.nonceSize)])
        let ciphertext = Array(encrypted[(8 + Shield.nonceSize)..<(encrypted.count - Shield.macSize)])
        let receivedMac = Array(encrypted.suffix(Shield.macSize))

        // Check counter
        if counter < recvCounter {
            throw ShieldError.replayDetected
        }
        if counter > recvCounter {
            throw ShieldError.outOfOrder
        }

        let messageKey = RatchetSession.deriveChainKey(recvKey, info: "message")

        // Verify MAC
        let macData = Array(encrypted.prefix(8)) + nonce + ciphertext
        let expectedMac = Array(Shield.hmacSha256(key: messageKey, data: macData).prefix(Shield.macSize))

        guard Shield.constantTimeEquals(receivedMac, expectedMac) else {
            throw ShieldError.authenticationFailed
        }

        // Decrypt
        let keystream = generateKeystream(key: messageKey, nonce: nonce, length: ciphertext.count)
        var plaintext = [UInt8](repeating: 0, count: ciphertext.count)
        for i in 0..<ciphertext.count {
            plaintext[i] = ciphertext[i] ^ keystream[i]
        }

        // Ratchet
        recvKey = RatchetSession.deriveChainKey(recvKey, info: "ratchet")
        recvCounter += 1

        return plaintext
    }

    public var currentSendCounter: UInt64 { sendCounter }
    public var currentRecvCounter: UInt64 { recvCounter }

    public func wipe() {
        Shield.secureWipe(&sendKey)
        Shield.secureWipe(&recvKey)
    }

    private static func deriveChainKey(_ key: [UInt8], info: String) -> [UInt8] {
        return Shield.sha256(key + Array(info.utf8))
    }

    private func generateKeystream(key: [UInt8], nonce: [UInt8], length: Int) -> [UInt8] {
        let numBlocks = (length + 31) / 32
        var keystream = [UInt8]()
        keystream.reserveCapacity(numBlocks * 32)

        for i in 0..<numBlocks {
            var block = key + nonce
            block.append(contentsOf: withUnsafeBytes(of: UInt32(i).littleEndian) { Array($0) })
            let hash = Shield.sha256(block)
            keystream.append(contentsOf: hash)
        }

        return Array(keystream.prefix(length))
    }
}

File: csharp/Shield/RatchetSession.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace Guard8.Shield
{
    /// <summary>
    /// RatchetSession provides forward secrecy through key ratcheting.
    /// </summary>
    public class RatchetSession : IDisposable
    {
        private byte[] _sendKey;
        private byte[] _recvKey;
        private ulong _sendCounter;
        private ulong _recvCounter;
        private readonly bool _isInitiator;
        private bool _disposed;

        public RatchetSession(byte[] rootKey, bool isInitiator)
        {
            if (rootKey.Length != Shield.KeySize)
                throw new ArgumentException("Invalid key size", nameof(rootKey));

            _isInitiator = isInitiator;
            _sendCounter = 0;
            _recvCounter = 0;

            if (isInitiator)
            {
                _sendKey = DeriveChainKey(rootKey, "init_send");
                _recvKey = DeriveChainKey(rootKey, "init_recv");
            }
            else
            {
                _sendKey = DeriveChainKey(rootKey, "init_recv");
                _recvKey = DeriveChainKey(rootKey, "init_send");
            }
        }

        public byte[] Encrypt(byte[] plaintext)
        {
            byte[] messageKey = DeriveChainKey(_sendKey, "message");
            byte[] nonce = Shield.RandomBytes(Shield.NonceSize);

            // Generate keystream and XOR
            byte[] keystream = GenerateKeystream(messageKey, nonce, plaintext.Length);
            byte[] ciphertext = new byte[plaintext.Length];
            for (int i = 0; i < plaintext.Length; i++)
                ciphertext[i] = (byte)(plaintext[i] ^ keystream[i]);

            // Counter bytes
            byte[] counterBytes = BitConverter.GetBytes(_sendCounter);

            // MAC over counter || nonce || ciphertext
            byte[] macData = new byte[8 + Shield.NonceSize + ciphertext.Length];
            Array.Copy(counterBytes, 0, macData, 0, 8);
            Array.Copy(nonce, 0, macData, 8, Shield.NonceSize);
            Array.Copy(ciphertext, 0, macData, 8 + Shield.NonceSize, ciphertext.Length);
            byte[] mac = Shield.HmacSha256(messageKey, macData);

            // Ratchet
            _sendKey = DeriveChainKey(_sendKey, "ratchet");
            _sendCounter++;

            // Format: counter(8) || nonce(16) || ciphertext || mac(16)
            byte[] result = new byte[8 + Shield.NonceSize + ciphertext.Length + Shield.MacSize];
            Array.Copy(counterBytes, 0, result, 0, 8);
            Array.Copy(nonce, 0, result, 8, Shield.NonceSize);
            Array.Copy(ciphertext, 0, result, 8 + Shield.NonceSize, ciphertext.Length);
            Array.Copy(mac, 0, result, 8 + Shield.NonceSize + ciphertext.Length, Shield.MacSize);

            Shield.SecureWipe(messageKey);
            return result;
        }

        public byte[] Decrypt(byte[] encrypted)
        {
            if (encrypted.Length < 8 + Shield.NonceSize + Shield.MacSize)
                throw new ArgumentException("Ciphertext too short");

            // Parse
            ulong counter = BitConverter.ToUInt64(encrypted, 0);
            byte[] nonce = new byte[Shield.NonceSize];
            Array.Copy(encrypted, 8, nonce, 0, Shield.NonceSize);

            int ciphertextLen = encrypted.Length - 8 - Shield.NonceSize - Shield.MacSize;
            byte[] ciphertext = new byte[ciphertextLen];
            Array.Copy(encrypted, 8 + Shield.NonceSize, ciphertext, 0, ciphertextLen);

            byte[] receivedMac = new byte[Shield.MacSize];
            Array.Copy(encrypted, encrypted.Length - Shield.MacSize, receivedMac, 0, Shield.MacSize);

            // Check counter
            if (counter < _recvCounter)
                throw new CryptographicException("Replay detected");
            if (counter > _recvCounter)
                throw new CryptographicException("Out of order message");

            byte[] messageKey = DeriveChainKey(_recvKey, "message");

            // Verify MAC
            byte[] macData = new byte[8 + Shield.NonceSize + ciphertextLen];
            Array.Copy(encrypted, 0, macData, 0, 8);
            Array.Copy(nonce, 0, macData, 8, Shield.NonceSize);
            Array.Copy(ciphertext, 0, macData, 8 + Shield.NonceSize, ciphertextLen);
            byte[] expectedMac = Shield.HmacSha256(messageKey, macData);

            if (!Shield.ConstantTimeEquals(receivedMac, expectedMac, Shield.MacSize))
            {
                Shield.SecureWipe(messageKey);
                throw new CryptographicException("Authentication failed");
            }

            // Decrypt
            byte[] keystream = GenerateKeystream(messageKey, nonce, ciphertextLen);
            byte[] plaintext = new byte[ciphertextLen];
            for (int i = 0; i < ciphertextLen; i++)
                plaintext[i] = (byte)(ciphertext[i] ^ keystream[i]);

            // Ratchet
            _recvKey = DeriveChainKey(_recvKey, "ratchet");
            _recvCounter++;

            Shield.SecureWipe(messageKey);
            return plaintext;
        }

        public ulong SendCounter => _sendCounter;
        public ulong RecvCounter => _recvCounter;

        private static byte[] DeriveChainKey(byte[] key, string info)
        {
            using var sha = SHA256.Create();
            byte[] data = new byte[key.Length + info.Length];
            Array.Copy(key, 0, data, 0, key.Length);
            Encoding.UTF8.GetBytes(info).CopyTo(data, key.Length);
            return sha.ComputeHash(data);
        }

        private static byte[] GenerateKeystream(byte[] key, byte[] nonce, int length)
        {
            int numBlocks = (length + 31) / 32;
            byte[] keystream = new byte[numBlocks * 32];

            for (int i = 0; i < numBlocks; i++)
            {
                byte[] block = new byte[Shield.KeySize + Shield.NonceSize + 4];
                Array.Copy(key, 0, block, 0, Shield.KeySize);
                Array.Copy(nonce, 0, block, Shield.KeySize, Shield.NonceSize);
                BitConverter.GetBytes(i).CopyTo(block, Shield.KeySize + Shield.NonceSize);

                byte[] hash = Shield.Sha256(block);
                Array.Copy(hash, 0, keystream, i * 32, 32);
            }

            byte[] result = new byte[length];
            Array.Copy(keystream, result, length);
            return result;
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                Shield.SecureWipe(_sendKey);
                Shield.SecureWipe(_recvKey);
                _disposed = true;
            }
        }
    }
}
